File: SMI-SDK/.NET SDK/Library/gis/trace/appender/SMI/SMILoggerAppender.cs
namespace pt.sapo.gis.trace.appender.SMI
{
    using System;
    using System.Linq;
    using System.Collections.Generic;
    using pt.sapo.sdb;
    using log4net.Core;
    using pt.sapo.sdb.smi;
    using pt.sapo.gis.trace.configuration;


    public abstract class SMILoggerAppender : Appender
    {
        public static readonly String REPORT_ACTIVITY_ID_PROPERTY = "activityId";
        public static readonly String REPORT_AGENT_ID_PROPERTY = "agentId";
        public static readonly String REPORT_SOURCE_ID_PROPERTY = "sourceId";
        public static readonly String FAILURE_SOURCE_ID_PROPERTY = "sourceId";

        private delegate bool EntryFilterHandler(pt.sapo.gis.trace.Entry e);

        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        public new SMILoggerAppenderConfigProperties Properties { get; protected set; }


        /// <summary>
        /// 
        /// </summary>
        /// <param name="config"></param>
        /// <exception cref="IntializatonException"></exception>
        public SMILoggerAppender(pt.sapo.gis.trace.configuration.appender config) : base(config) {
            this.Properties = new SMILoggerAppenderConfigProperties(base.Properties);
            base.Properties = this.Properties;
        }

        private String GetFailureCodeFromEntry(pt.sapo.gis.trace.Entry e)
        {
            return e.Type.ToString();            
        }

        public override void OnEntry(trace.Entry e)
        {
            var failures = new List<Failure>();
            FillFailure(e, failures, et => severities.ContainsKey(eventType.entry) == false || severities[eventType.entry].Contains(et.Severity));
            ManagementReport report = BuildReport(e.Parent, failures);
            Send(report, e.Parent.Properties.ContainsKey(REPORT_ACTIVITY_ID_PROPERTY) ? e.Parent.Properties[REPORT_ACTIVITY_ID_PROPERTY] as String : null, e.Parent.Properties.ContainsKey(REPORT_AGENT_ID_PROPERTY) ? e.Parent.Properties[REPORT_AGENT_ID_PROPERTY] as String : null);
        }

        public override void OnTrace(trace.Trace t)
        {
            var failures = new List<Failure>();
            foreach (trace.Entry e in t.Entries)
                FillFailures(e, failures, et => severities.ContainsKey(eventType.trace) == false || severities[eventType.trace].Contains(et.Severity));

            ManagementReport report = BuildReport(t, failures.ToArray());
            Send(report, t.Properties.ContainsKey(REPORT_ACTIVITY_ID_PROPERTY) ? t.Properties[REPORT_ACTIVITY_ID_PROPERTY] as String : null, t.Properties.ContainsKey(REPORT_AGENT_ID_PROPERTY) ? t.Properties[REPORT_AGENT_ID_PROPERTY] as String : null);
        }

        private ManagementReport BuildReport(trace.Trace trace, IEnumerable<Failure> failures)
        {
            ManagementReport report = new ManagementReport
            {
                ID = trace.Id.ToString(),
                State = new State()
                {
                    Execution = ExecutionState.Active,
                    Health = HealthState.OperationalWithFailures,
                    Failures = failures.ToArray()
                },
                DateTime = DateTime.UtcNow,
                SourceID = trace.Properties.ContainsKey(REPORT_SOURCE_ID_PROPERTY) ? trace.Properties[REPORT_SOURCE_ID_PROPERTY] as String : Properties.DefaultSourceId
            };

            return report;            
        }

        protected abstract void Send(ManagementReport report, String activityId, String agentId);

        private void FillFailure(pt.sapo.gis.trace.Entry entry, List<Failure> failures, EntryFilterHandler filter)
        {
            // only add entry if there is no filter or it pass the filter test
            if (filter(entry))
            {
                Failure failure = new Failure
                {
                    FailureID = GetFailureCodeFromEntry(entry),
                    SourceID = entry.Properties.ContainsKey(FAILURE_SOURCE_ID_PROPERTY) ? entry.Properties[FAILURE_SOURCE_ID_PROPERTY] as String : (TraceManager.Trace.Properties.ContainsKey(REPORT_SOURCE_ID_PROPERTY) ? TraceManager.Trace.Properties[REPORT_SOURCE_ID_PROPERTY] as String : Properties.DefaultSourceId),
                    Details = entry.Properties
                };
                failures.Add(failure);
            }
        }

        private void FillFailures(trace.Entry entry, List<Failure> failures, EntryFilterHandler filter)
        {
            FillFailure(entry, failures, filter);
            // fill child entries
            foreach (trace.Entry innerEntry in entry.Entries)
                FillFailures(innerEntry, failures, filter);
        }        
    }
}

File: SMI-SDK/Java SDK/src/pt/sapo/gis/trace/appender/SMI/SMILoggerAppender.java
package pt.sapo.gis.trace.appender.SMI;

import java.util.Date;
import java.util.EnumMap;
import java.util.GregorianCalendar;
import java.util.TimeZone;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.xml.datatype.DatatypeConfigurationException;
import javax.xml.datatype.DatatypeFactory;
import javax.xml.datatype.XMLGregorianCalendar;
import org.tmforum.schemas.smi.ExecutionState;
import org.tmforum.schemas.smi.Failure;
import org.tmforum.schemas.smi.Failures;
import org.tmforum.schemas.smi.HealthState;
import org.tmforum.schemas.smi.ManagementReport;
import org.tmforum.schemas.smi.State;

import pt.sapo.gis.trace.Entry;
import pt.sapo.gis.trace.EntryType;

import pt.sapo.gis.trace.SeverityEventFilter;
import pt.sapo.gis.trace.Trace;
import pt.sapo.gis.trace.appender.Appender;
import pt.sapo.sdb.trace.FailureCode;

public abstract class SMILoggerAppender extends Appender{
    public static final String REPORT_ACTIVITY_ID_PROPERTY = "activityId";
    public static final String REPORT_SOURCE_ID_PROPERTY = "sourceId";
    public static final String FAILURE_SOURCE_ID_PROPERTY = "sourceId";

    private static final EnumMap<EntryType, FailureCode> codeMapping = new EnumMap<EntryType, FailureCode>(EntryType.class);
        
    protected SMILoggerAppenderConfigProperties properties;
    
    public SMILoggerAppender(pt.sapo.gis.trace.configuration.Appender config){
        super(config);
        properties = new SMILoggerAppenderConfigProperties(config.getProperties());
    }
    
    @Override
    protected void entryEventHandler(Entry e) {
        Failures failures = new Failures();
        FillFailure(e, failures, entrySeverityFilter);
        
        ManagementReport report = BuildReport(e.getParent(), failures);
        send(report);
    }

    @Override
    protected void traceEventHandler(Trace t) {        
        Failures failures = new Failures();
        for(Entry e : t.getEntries())
            FillFailures(e, failures, traceSeverityFilter);            
        
        ManagementReport report = BuildReport(t, failures);
        send(report);
    }
    
    private ManagementReport BuildReport(Trace trace, Failures failures){
        ManagementReport report = new ManagementReport();
        
        report.setSourceID(getReportSourceId(trace));
        report.setID(trace.getId());
        
        //report.setActivityID = (String)t.getProperty(REPORT_ACTIVITY_ID_PROPERTY)
        
        State state = new State();
        state.setExecution(ExecutionState.ACTIVE);
        state.setHealth(HealthState.OPERATIONAL_WITH_FAILURES);
        state.setFailures(failures);
        report.setState(state);
        
        GregorianCalendar cal = new GregorianCalendar();
        cal.setTime(new Date());
        cal.setTimeZone(TimeZone.getTimeZone("GMT"));
        XMLGregorianCalendar dateTime = null;
        try {
            dateTime = DatatypeFactory.newInstance().newXMLGregorianCalendar(cal);
        } catch (DatatypeConfigurationException ex) {
            Logger.getLogger(SMILoggerAppender.class.getName()).log(Level.SEVERE, null, ex);
        }
        
        report.setDateTime(dateTime);
        //report.setMetrics(null);
        
        return report;
    }

    private String GetFailureIdFromEntry(Entry e){
        if(e.getType() != null){            
            return e.getType().toString();
        }

        return FailureCode.UNKNOWN.name();
    }
    
    private void FillFailure(Entry entry, Failures  failures, SeverityEventFilter filter) {
        // only add entry if there is no filter or it pass the filter test
        if(filter == null || filter.contains(entry.getSeverity())){
            Failure failure = new Failure();
            failure.setFailureID(GetFailureIdFromEntry(entry));
            
            failure.setSourceID(getFailureSourceId(entry));
            if(entry.getProperties() != null && !entry.getProperties().isEmpty())
                failure.setDetails(entry.getProperties());
            failures.getFailure().add(failure);
        }
    }
    
    private void FillFailures(Entry entry, Failures failures, SeverityEventFilter filter) {
        FillFailure(entry, failures, filter);
        // fill child entries
        for(Entry innerEntry : entry.getEntries())
            FillFailures(innerEntry, failures, filter);
    }

    private String getReportSourceId(Trace t){
        Object sourceIdProp = t.getProperty(REPORT_SOURCE_ID_PROPERTY);
        if(sourceIdProp != null)       
            return (String)sourceIdProp;
        return properties.getDefaultSourceID();
    }
    
    private String getFailureSourceId(Entry e){
        Object sourceIdProp = e.getProperty(FAILURE_SOURCE_ID_PROPERTY);
        if(sourceIdProp != null)       
            return (String)sourceIdProp;
        return getReportSourceId(e.getParent());
    }
    
    protected abstract void send(ManagementReport report);   
}